### PR TITLE
Fix test rules functionality 

### LINKED ISSUE #445: Test rules functionality is unstable
### Description
There seems to be faulty logic in the test rules functionality. If there are more events than there are rules it will not aggregate. More events causes a json index issue.

### Motivation
The test rules functionality does not currently work as expected.

### Exemplification
    2020-02-18 15:47:48.234 ERROR 27304 --- [nio-8090-exec-4] c.e.e.controller.RuleTestControllerImpl  : Internal Server Error: Failed to generate aggregated object.

    org.json.JSONException: JSONArray[7] not found.
	at org.json.JSONArray.get(JSONArray.java:218) ~[json-20180130.jar:na]
	at org.json.JSONArray.getJSONObject(JSONArray.java:423) ~[json-20180130.jar:na]
	at com.ericsson.ei.services.RuleCheckService.prepareAggregatedObject(RuleCheckService.java:45) ~[classes/:na]
	at com.ericsson.ei.controller.RuleTestControllerImpl.createRuleTestRunFullAggregation(RuleTestControllerImpl.java:90) ~[classes/:na]
	at com.ericsson.ei.controller.RuleTestControllerImpl$$FastClassBySpringCGLIB$$8aca890d.invoke(<generated>) [classes/:na]

### Benefits
A working test rules page.

### Possible Drawbacks
None.

### PR DESCRIPTION
**Pull request by e-pettersson-ericsson**
### Applicable Issues
Fixes #445 and #448 

### Description of the Change
#445 was caused due to a flaw in logic in RulesTestService.prepareAggregatedObject method. The iteration of list of rules made it impossible to have more events than rules in the test aggregation. 

#448 was caused because a check was never made on the template names in the list of rules (the code block was always skipped) Made sure to handle this with an InvalidRulesException beign thrown.

I also made a lot of refactoring, including:
* renaming all "ruleCheck" to "ruleTest"
* moved RuleTestService + IRulesTestService from service package to rules package (because they have also changed names Git thinks they are new files...)
* split up prepareAggregatedObject into several methods (and added javadoc)
* added an extra event in the existing cucumber tests to ensure #445 was fixed.
* add extra test to cover the fix for #448 
* updated cucumber scenario of testing invalid rules to actually have invalid rules
* simplified the events and rules used in tests 
* refactored TestRulesRestAPI and moved out reading test resources to a Before method.
* Improved error messages in Controller for TestRules
* Handle extra error case when rules are invalid

### Benefits
Fully functional test rules page which can handle more events (i.e. 3 rules and 5 events are no longer a problem, but will be aggregated properly)
If a user has included different template names in their rules to be tested, the rules will fail validation
and users are encouraged to fix before trying again.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Emelie Pettersson emelie.pettersson@ericsson.com

### CODE REVIEW

File: src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java [hunk 1]
```diff
@@ -21,10 +21,13 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 import java.io.File;
+import java.io.IOException;
 
+import net.bytebuddy.implementation.bytecode.Throw;
```
Comment (e-pettersson-ericsson) [inline src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java]: Codacy Issue found: [Avoid unused imports such as 'net.bytebuddy.implementation.bytecode.Throw'](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=5208134)

File: src/main/java/com/ericsson/ei/rules/IRuleTestService.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.ericsson.ei.rules;
+
+import java.io.IOException;
+
+import com.ericsson.ei.exception.InvalidRulesException;
+import org.json.JSONArray;
+import org.json.JSONException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/rules/IRuleTestService.java]: Codacy Issue found: [Unused import - com.fasterxml.jackson.core.JsonProcessingException.](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=5208134)

File: src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java (hunk 1 above)
Comment (e-pettersson-ericsson) [inline src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java]: Codacy Issue found: [Unused import - net.bytebuddy.implementation.bytecode.Throw.](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=5208134)

File: src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java [hunk 3]
```diff
@@ -63,35 +65,46 @@
...
-    private static final Logger LOGGER = LoggerFactory.getLogger(TestRulesRestAPI.class);
-    private static final String INPUT_FILE_PATH = "src/test/resources/EiffelArtifactCreatedEvent.json";
-    private static final String EXTRACTION_RULE_FILE_PATH = "src/test/resources/ExtractionRule.txt";
+    private static final String SINGLE_EVENT = "src/test/resources/EiffelArtifactCreatedEvent.json";
+    private static final String SINGLE_EXTRACTION_RULE = "src/test/resources/ExtractionRule.txt";
     private static final String EVENTS = "src/test/resources/AggregateListEvents.json";
     private static final String RULES = "src/test/resources/AggregateListRules.json";
     private static final String AGGREGATED_RESULT_OBJECT = "src/test/resources/AggregateResultObject.json";
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(TestRulesRestAPI.class);
```
Comment (e-pettersson-ericsson) [inline src/test/java/com/ericsson/ei/rules/TestRulesRestAPI.java]: Codacy Issue found: [Avoid unused private fields such as 'LOGGER'.](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=5208134)

File: src/main/java/com/ericsson/ei/rules/IRuleTestService.java (hunk 2 above)
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/rules/IRuleTestService.java]: Codacy Issue found: [Avoid unused imports such as 'com.fasterxml.jackson.core.JsonProcessingException'](https://app.codacy.com/manual/eiffel-intelligence-maintainers/eiffel-intelligence/pullRequest?prid=5208134)

File: src/main/java/com/ericsson/ei/rules/RuleTestService.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        eventToObjectMapHandler.deleteEventObjectMap(templateName);
+
+        return response;
+    }
+
+    /**
+     * This method iterates through the list of rules to extract the template name from each rule.
+     * All rules should contain the same template name, which is returned.
+     *
+     * @throws  InvalidRulesException if the rules contains several template names
```
Comment (Christoffer-Cortes) [inline src/main/java/com/ericsson/ei/rules/RuleTestService.java]: Probably very nit picky but "contains" should be "contain". =P

Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/rules/RuleTestService.java]: Thanks, I missed that!

**Pull request merged**